designs/2019-expose-rules-to-formatters/readme.MD

```diff
@@ -0,0 +1,137 @@
+
+The [Working with Custom Formatters](https://eslint.org/docs/developer-guide/working-with-custom-formatters) article will have to be updated:
+* Code samples will need the new `data` parameter added wherever the exported interface function is shown, *but only when it is used*.
+* The `data` parameter should be called out and described, and should include a link to the [Working with Rules](https://eslint.org/docs/developer-guide/working-with-rules) article. The primary goal here is to familiarize formatter author with the structure of the `data` parameter and Rule object.
+* It should be noted that Rule metadata properties such as description, category, and help URL are not required and may not be defined, and that custom formatter code should take this into account.
+* We should show the use of rule data in one of the examples by either modifying an existing one (maybe the [Detailed formatter](https://eslint.org/docs/developer-guide/working-with-custom-formatters#detailed-formatter) example) or adding a new one. One idea would be to suffix the results output with a list of rules that were violated, using a helper function something like this:
+
+```js
+var rulesViolated = {};
+...
+function printRules() {
+    var lines = "*** RULES:\n";
+    rulesViolated.forEach(function (rule) {
+        lines += rule.ruleId;
+        
+        if (rule.meta.docs.description) {
+            lines += ": " + rule.meta.docs.description;
+        }
+        
+        lines += "\n";
+        
+        if (rule.meta.docs.url) {
+            lines += rule.meta.docs.url + "\n";
+        }
+    });
+    return lines;
+}
+```
+
+## Drawbacks
+
+This is a fairly innocuous change in that it is additive, non-breaking, and does not change any of ESLint's core functionality. A downside is that we will be exposing the Rule data model to third-party developers, so future changes could break existing formatters. For example, removing or renaming an existing property, or changing the structure of the Rule object.
+
+## Backwards Compatibility Analysis
+
+Since this change is manifested as a new parameter to the formatter's exported interface function, existing formatter code will not be affected and will continue to function even without adding the new parameter to their exported function. I have confirmed this assertion via manual testing.
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+* Including the rule metadata in the result object. This approach results in redundant data being returned, and includes external metadata properties that are not directly relevant. (As an analogy, it's like ...)
```

Comment:
> (As an analogy, it's like ...)

I'm confused about what this means. Did you forget to finish this sentence?

Reply:
Sorry, meant to remove this.